sql/core/src/test/scala/org/apache/spark/sql/CollationExpressionWalkerSuite.scala

```diff
@@ -0,0 +1,592 @@
+        }
+      }
+
+      val exceptionLcase = {
+        try {
+          instanceLcase.eval(EmptyRow)
+          None
+        } catch {
+          case e: Throwable => Some(e)
+        }
+      }
+
+      // Check that both cases either throw or pass
+      assert(exceptionUtfBinary.isDefined == exceptionLcase.isDefined)
+
+      if (exceptionUtfBinary.isEmpty) {
+        val resUtf8Binary = instanceUtf8Binary.eval(EmptyRow)
+        val resUtf8Lcase = instanceLcase.eval(EmptyRow)
+
+        val dt = instanceLcase.dataType
+
+        dt match {
+          case _: StringType if resUtf8Lcase != null && resUtf8Lcase != null =>
+            assert(resUtf8Binary.isInstanceOf[UTF8String])
+            assert(resUtf8Lcase.isInstanceOf[UTF8String])
+            // scalastyle:off caselocale
+            assert(resUtf8Binary.asInstanceOf[UTF8String].toLowerCase.binaryEquals(
+              resUtf8Lcase.asInstanceOf[UTF8String].toLowerCase))
+          // scalastyle:on caselocale
+          case _ => resUtf8Lcase === resUtf8Binary
+        }
+      }
+      else {
+        assert(exceptionUtfBinary.get.getClass == exceptionLcase.get.getClass)
+      }
+    }
+  }
+
+  test("SPARK-48280: Expression Walker for codeGen generation") {
+
+    var (funInfos, toSkip) = extractRelevantExpressions()
+    toSkip = toSkip ++ List(
+      // Problem caught with other tests already
+      "map_from_arrays",
+      // These expressions are not called as functions
+      "lead",
+      "nth_value",
+      // Failing asserts
+      "session_window",
+      "ascii",
+      "to_xml"
+    )
+    for (f <- funInfos.filter(f => !toSkip.contains(f.getName))) {
+      println("checking - " + f.getName)
+      val cl = Utils.classForName(f.getClassName)
+      val headConstructor = cl.getConstructors.head
+      val params = headConstructor.getParameters.map(p => p.getType)
+      val args = generateData(params.toSeq, Utf8Binary)
+      val expr = headConstructor.newInstance(args: _*).asInstanceOf[ExpectsInputTypes]
+      val inputTypes = expr.inputTypes
+
+      withTable("tbl", "tbl_lcase") {
+        sql("CREATE TABLE tbl (" +
+          inputTypes.zipWithIndex
+            .map(it => "col" +
+              it._2.toString + " " +
+              generateInputTypeAsStrings(it._1, Utf8Binary)).mkString(", ") +
+          ") USING PARQUET")
+        sql("INSERT INTO tbl VALUES (" +
+          inputTypes.map(generateInputAsString(_, Utf8Binary)).mkString(", ") +
+          ")")
+
+        sql("CREATE TABLE tbl_lcase (" +
+          inputTypes.zipWithIndex
+            .map(it => "col" +
+              it._2.toString + " " +
+              generateInputTypeAsStrings(it._1, Utf8BinaryLcase)).mkString(", ") +
+          ") USING PARQUET")
+        sql("INSERT INTO tbl_lcase VALUES (" +
+          inputTypes.map(generateInputAsString(_, Utf8BinaryLcase)).mkString(", ") +
+          ")")
+
+        val utf8BinaryResult = try {
+          if (expr.isInstanceOf[BinaryComparison]) {
```

Comment:
why do we need to use SQL api here? Wouldn't using dataframe do the same job without manually crafting queries as sql strings?